### PR TITLE
preselected binaries

### PR DESCRIPTION
**Pull request by drhideg**
Implement selectable tasmota builds based on `BUILDS.md`.

- At the moment development branch can't compile because of a failed formatting language files in tasmota source (https://github.com/arendst/Tasmota/commit/59992d195463cc79cb39c667fdb9cb1f6106467a).
- I don't know how can i fire select event programatically to let the TasmoCompiler build version use the undefined features (-91k binary size)
- Sensors build version failed to compile with 2.7.4 core (.text1 not fit in region iram)
- IR version selects both IR base and full however it can't compile because of the used build flags
- There is something with the TuyaMCU feature, where it doesn't matter which binary version is the base to select as additional
- ...and maybe lot of other

These problems are beyond my knowledge, will anyone be able to deal with it?

### CODE REVIEW

**Comment by benzino77** [comment]
A lot of changes ;)
I will find probably more time next year to review this PR.

**Comment by benzino77** [comment]
I'm not sure it is a good direction. One could build predefined binaries types by using "Custom parameters" step. It can be accomplished by putting appropriate `#define` statement:
```
//#define FIRMWARE_LITE                            // Create tasmota-lite with no sensors
//#define FIRMWARE_SENSORS                         // Create tasmota-sensors with useful sensors enabled
//#define FIRMWARE_KNX_NO_EMULATION                // Create tasmota-knx with KNX but without Emulation
//#define FIRMWARE_DISPLAYS                        // Create tasmota-display with display drivers enabled
//#define FIRMWARE_IR                              // Create tasmota-ir with IR full protocols activated, and many sensors disabled
//#define FIRMWARE_IR_CUSTOM                       // Create tasmota customizable with special marker to add all IR protocols
//#define FIRMWARE_MINIMAL                         // Create tasmota-minimal as intermediate firmware for OTA-MAGIC
```

This means that one could check KNX feature on "Features" step and then put `#define FIRMWARE_SENSORS` in "Custom parameters" to build Sensors binary with KNX support.

**Comment by drhideg** [comment]
Thank you for dealing with it. Don't want to make it selectable before selecting features?
See https://github.com/benzino77/tasmocompiler/issues/110

**Comment by benzino77** [comment]
This is just one line in Custom parameters field. It will be waste of your energy to implement it as a GUI element or additional step. It' is better to put this energy to... update documentation to explain how to do this by just one line (but it has to be tested). 
I'm not saying that's unneeded but it can be accomplished by simpler solution (simpler also in terms of TasmoCompiler code).

**Comment by drhideg** [comment]
I agree with you that it can be too complicated, but when I look at myself, no one reads the documents :)
Especially with my English, you don’t want me to write a document.
In my opinion, the base binary should be specified before the features. It can be in the same step with a select box
Close this and i will try to come up with a simpler solution.

**Comment by benzino77** [comment]
We can communicate very fluent - do not exaggerate. 
It can be placed as a `Select` component in the "select features" step. It can has some options to choose:
- none (default)
- sensors
- displays
- lite
- ir

with appropriate description. Something like: "the binary will be based on <Tasmota Firmware version> plus selected features"

You have to assure (I think) that `#define FIRMWARE_*` is the last entry in `user_config_override.h`:

```
<...snip...>
#ifdef USE_DISPLAY_SSD1351
  #undef USE_DISPLAY_SSD1351
#endif

#ifdef USE_DISPLAY_RA8876
  #undef USE_DISPLAY_RA8876
#endif

#ifdef USE_ADC_VCC
  #undef USE_ADC_VCC
#endif

#ifdef USE_ZIGBEE
  #undef USE_ZIGBEE
#endif

#ifdef MY_LANGUAGE
  #undef MY_LANGUAGE
#endif
#define MY_LANGUAGE     pl_PL

#define FIRMWARE_SENSORS
#endif
```

**Comment by benzino77** [comment]
...putting `#define FIRMWARE_SENSORS` in "custom parameters" field will do the same and is much, much simpler than implementing it as a part of the GUI.

**Comment by drhideg** [comment]
The good thing is that you are interested in solving a problem itself :)

In the PR the lot of magic with the USE flags is there to see which Feature added to that base image.

There are plenty of non-working settings which i found, like `#define FIRMWARE_ZBBRIDGE` alway remove a lot of feature and you can not re-add  it in the "custom parameters" field, Like `#define USE_ADE7953` and `#define USE_ENERGY_SENSOR`. It will be removed by `tasmota_configurations.h`. At least I did not found any ADE related string in the map file while there was if
 I choose "Energy sensors" in "Feature step".

in short, I think it is not generally applicable.

**Comment by benzino77** [comment]
So my first thought was correct - it's better to assure all sensors coverage in the "features step" and update appropriate documentation then implementing this as a GUI feature. In this case one could choose the "checkbox" with sensor or sensors she/he is interested in. Sensors coverage can be checked in the [FEATURES_DESC ](https://github.com/benzino77/tasmocompiler/blob/master/FEATURES_DESC.md) file (which has to be updated). If there is no sensor which is needed then it can be included in the binary by "Custom parameters" text field.

**Comment by Jason2866** [comment]
You cant (easily) add features to predefined Tasmota variants (as designed and wanted!)
the settings done in `user_config_override.h` will be overwritten from the hardcoded settings for this build. This is done in file https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_configurations.h and https://github.com/arendst/Tasmota/blob/development/tasmota/tasmota_configurations_ESP32.h You can use this file for information what is "in" a predefined build.
Since a few days Tasmota supports now officially builds >1 MByte. Be aware updating not supported builds bigger 1Mbyte will loose settings and maybe needs serial flashing.
The ZigbeeBridge build is now fixed compiled with a 2M linker script!!
Flashing with a different Linker script (the standard 1M one too!) will result in data loss
